### PR TITLE
Transfer filesystem utility to jMonocle

### LINKED ISSUE #7: Add FileSystemUtility to this repository
Need to port the changes in this PR to libs/filesystem in this repository.
https://github.com/padaiyal/jFileSystem/pull/3/files

Make sure to update the README in libs/filesystem as well.

### PR DESCRIPTION
**Pull request by abprash**
## Description
Port the Filesystem utility over to jMonocle.
Source PR - https://github.com/padaiyal/jFileSystem/pull/3/files

Fixes #7.

## Author's Checklist
- [x] **These changes don't break existing dependants.** If they do, and you suggest to proceed, 
please explain why.

## Reviewer's Checklist
- [x] **All classes have documentation comments.**
- [x] **All methods have documentation comments.**
- [ ] **No parameters are hardcoded.** Should be loaded as properties instead. If there needs to be hardcoded properties, explain why.
- [x] **All information to be logged/displayed to the user are internationalized.** If not, explain why. 
- [x] **README.md has been updated if needed.**
- [x] **pom.xml version is set to the latest date** If not, explain why.

## Assignee's Checklist
- [x] **All GitHub action checks have passed.**
- [x] **All reviewers have approved.**
- [x] **Relevant documentation has been updated if needed.**
- [x] **pom.xml version is set to the latest date** If not, explain why.

### CODE REVIEW

File: libs/filesystem/src/main/java/org/padaiyal/utilities/filesystem/FileSystemUtility.java
Comment (abprash) [reply libs/filesystem/src/main/java/org/padaiyal/utilities/filesystem/FileSystemUtility.java:492]: @sonatype-lift ignore

Comment (abprash) [reply libs/filesystem/src/main/java/org/padaiyal/utilities/filesystem/FileSystemUtility.java:492]: ignoring for now - as there's no benefit to changing to list for this scenario.

File: libs/filesystem/src/main/java/org/padaiyal/utilities/filesystem/FileSystemUtility.java
Comment (abprash) [reply libs/filesystem/src/main/java/org/padaiyal/utilities/filesystem/FileSystemUtility.java]: looks like a false +ve to me. I can't see how a user input can be used here.
@the-c0d3br34k3r to weigh in.

File: libs/filesystem/src/main/java/org/padaiyal/utilities/filesystem/FileSystemUtility.java
Comment (abprash) [reply libs/filesystem/src/main/java/org/padaiyal/utilities/filesystem/FileSystemUtility.java]: @sonatype-lift ignore

Comment (abprash) [reply libs/filesystem/src/main/java/org/padaiyal/utilities/filesystem/FileSystemUtility.java]: @sonatype-lift ignore

File: libs/filesystem/pom.xml
Comment (the-c0d3br34k3r) [inline libs/filesystem/pom.xml:18]: - [ ] FYI, the <version> tag for this pom.xml, the one in libs and in the root folder have to be changed to the date when the PR is approved.

File: libs/filesystem/src/main/java/org/padaiyal/utilities/filesystem/wordspackage/WordCounter.java
Comment (the-c0d3br34k3r) [inline libs/filesystem/src/main/java/org/padaiyal/utilities/filesystem/wordspackage/WordCounter.java:1]: The words package and it's corresponding tests were template contents and need to be removed.

File: libs/filesystem/src/main/java/org/padaiyal/utilities/filesystem/parameterconverters/ExceptionClassConverter.java
Comment (the-c0d3br34k3r) [inline libs/filesystem/src/main/java/org/padaiyal/utilities/filesystem/parameterconverters/ExceptionClassConverter.java:1]: Refer to the unittestextras package which has this class. This file can be removed.

File: libs/properties/src/main/java/org/padaiyal/utilities/PropertyUtility.java
Comment (the-c0d3br34k3r) [inline libs/properties/src/main/java/org/padaiyal/utilities/PropertyUtility.java:58]: What was the motivation for this change in the properties package?

File: libs/filesystem/pom.xml
Comment (the-c0d3br34k3r) [inline libs/filesystem/pom.xml:21]: Ensure that the pom.xml in libs has a module entry for filesystem.

File: libs/filesystem/src/main/java/org/padaiyal/utilities/filesystem/FileSystemUtility.java
Comment (the-c0d3br34k3r) [inline libs/filesystem/src/main/java/org/padaiyal/utilities/filesystem/FileSystemUtility.java:119]: Try running the auto linter in your IDE with the google check style config.
https://github.com/google/google-java-format#intellij-android-studio-and-other-jetbrains-ides

File: libs/filesystem/src/test/java/tests/org/padaiyal/utilities/testutils/ExceptionClassConverter.java
Comment (the-c0d3br34k3r) [inline libs/filesystem/src/test/java/tests/org/padaiyal/utilities/testutils/ExceptionClassConverter.java:1]: This is present in the unittestextras package which can be added as a dependency and then this file can be removed.

File: libs/filesystem/src/test/java/tests/org/padaiyal/utilities/testutils/StringArrayConverter.java
Comment (the-c0d3br34k3r) [inline libs/filesystem/src/test/java/tests/org/padaiyal/utilities/testutils/StringArrayConverter.java:1]: This is present in the unittestextras package which can be added as a dependency and then this file can be removed.

File: libs/filesystem/src/main/java/org/padaiyal/utilities/filesystem/FileSystemUtility.java
Comment (the-c0d3br34k3r) [reply libs/filesystem/src/main/java/org/padaiyal/utilities/filesystem/FileSystemUtility.java]: Yes, looks like a false positive.

@sonatype-lift ignore

File: libs/filesystem/src/main/java/org/padaiyal/utilities/filesystem/FileSystemUtility.java
Comment (the-c0d3br34k3r) [reply libs/filesystem/src/main/java/org/padaiyal/utilities/filesystem/FileSystemUtility.java:638]: @sonatype-lift ignore

File: libs/filesystem/src/main/java/org/padaiyal/utilities/filesystem/FileSystemUtility.java
Comment (the-c0d3br34k3r) [reply libs/filesystem/src/main/java/org/padaiyal/utilities/filesystem/FileSystemUtility.java]: @sonatype-lift ignore

File: libs/filesystem/src/main/java/org/padaiyal/utilities/filesystem/FileSystemUtility.java
Comment (the-c0d3br34k3r) [reply libs/filesystem/src/main/java/org/padaiyal/utilities/filesystem/FileSystemUtility.java]: @abprash we would want to initialize `Files.list(path)` within a try-with-resources block. Looks like it's a valid leak.
https://errorprone.info/bugpattern/StreamResourceLeak

There seems to be another occurrence of this as well.

File: libs/filesystem/src/main/java/org/padaiyal/utilities/filesystem/FileSystemUtility.java
Comment (the-c0d3br34k3r) [inline libs/filesystem/src/main/java/org/padaiyal/utilities/filesystem/FileSystemUtility.java:96]: I'm presuming you added this if clause to fix this [issue](https://github.com/padaiyal/jMonocle/pull/11#discussion_r1118310444) ?
We'd need to either log a warning/error or raise an exception when `subPathWatchServiceInfo` is null.

**Review by the-c0d3br34k3r — CHANGES_REQUESTED** [review]
Second round.

File: libs/properties/src/main/java/org/padaiyal/utilities/PropertyUtility.java
Comment (abprash) [reply libs/properties/src/main/java/org/padaiyal/utilities/PropertyUtility.java:58]: unit test was failing without this.
By default, it throws InvalidArgumentException, but the junit was expecting NoSuchElementException.

File: libs/filesystem/src/main/java/org/padaiyal/utilities/filesystem/FileSystemUtility.java
Comment (abprash) [reply libs/filesystem/src/main/java/org/padaiyal/utilities/filesystem/FileSystemUtility.java:119]: Didn't notice this, was in the original PR.

**Comment by abprash** [comment]
> Thanks @abprash ! This is a huge port and a good value add to this repo.
> 
> I've added comments requesting a few changes. And in the CI, FileSystemUtilityTest never ran as filesystem has to be a module entry in libs/pom.xml.
> 
> Also update the filesystem README to reflect the current state. Don't use the existing README as a template. Refer the READMEs in properties or i18n.

Sure, will address the comments. Thanks @the-c0d3br34k3r !

File: libs/properties/src/main/java/org/padaiyal/utilities/PropertyUtility.java
Comment (the-c0d3br34k3r) [reply libs/properties/src/main/java/org/padaiyal/utilities/PropertyUtility.java:58]: Which test was failing?
If it's in filesystem, then it should be changed to expect an InvalidArgumentException.
If it's in properties, then there should be a pre-existing failure which wasn't the case.

File: libs/filesystem/src/main/java/org/padaiyal/utilities/filesystem/wordspackage/WordCounter.java
Comment (abprash) [reply libs/filesystem/src/main/java/org/padaiyal/utilities/filesystem/wordspackage/WordCounter.java:1]: good, will remove it then.

**Comment by abprash** [comment]
>And in the CI, FileSystemUtilityTest never ran as filesystem has to be a module entry in libs/pom.xml.
Can you clarify what you meant by this ? I see filesystem is already a module in libs/pom.xml

File: libs/filesystem/pom.xml
Comment (abprash) [reply libs/filesystem/pom.xml:21]: Yes, it already did!

File: libs/properties/src/main/java/org/padaiyal/utilities/PropertyUtility.java
Comment (abprash) [reply libs/properties/src/main/java/org/padaiyal/utilities/PropertyUtility.java:58]: it's a junit in properties. Not in filesystem.
Let me share the log from the github actions, the next time i see it.

File: libs/filesystem/src/main/java/org/padaiyal/utilities/filesystem/FileSystemUtility.java
Comment (abprash) [reply libs/filesystem/src/main/java/org/padaiyal/utilities/filesystem/FileSystemUtility.java:96]: which one would be preferable ? do we prefer that it fail or is it more of a warning ?

File: libs/filesystem/src/main/java/org/padaiyal/utilities/filesystem/FileSystemUtility.java
Comment (abprash) [reply libs/filesystem/src/main/java/org/padaiyal/utilities/filesystem/FileSystemUtility.java]: have already addressed this one. Let me check the other one. Likely not in the diff.

File: libs/filesystem/src/main/java/org/padaiyal/utilities/filesystem/FileSystemUtility.java
Comment (abprash) [reply libs/filesystem/src/main/java/org/padaiyal/utilities/filesystem/FileSystemUtility.java]: fixed

File: libs/filesystem/src/main/java/org/padaiyal/utilities/filesystem/FileSystemUtility.java
Comment (abprash) [reply libs/filesystem/src/main/java/org/padaiyal/utilities/filesystem/FileSystemUtility.java:505]: @sonatype-lift ignore

File: libs/filesystem/src/main/java/org/padaiyal/utilities/filesystem/FileSystemUtility.java
Comment (the-c0d3br34k3r) [reply libs/filesystem/src/main/java/org/padaiyal/utilities/filesystem/FileSystemUtility.java:96]: I would vote to log a warning message with subPath and pathToWatchServiceRegistrationInfoMapCopy.
And skip the current iteration and continue.

Comment (abprash) [reply libs/filesystem/src/main/java/org/padaiyal/utilities/filesystem/FileSystemUtility.java:96]: sounds good. Will add a warning.

File: libs/filesystem/src/main/java/org/padaiyal/utilities/filesystem/FileSystemUtility.java
Comment (abprash) [reply libs/filesystem/src/main/java/org/padaiyal/utilities/filesystem/FileSystemUtility.java]: Have addressed the remaining ones as well.

File: libs/filesystem/src/main/java/org/padaiyal/utilities/filesystem/FileSystemUtility.java
Comment (abprash) [reply libs/filesystem/src/main/java/org/padaiyal/utilities/filesystem/FileSystemUtility.java:303]: Had refactored to address stream leak.

Comment (abprash) [reply libs/filesystem/src/main/java/org/padaiyal/utilities/filesystem/FileSystemUtility.java:303]: @sonatype-lift ignore

File: libs/filesystem/src/main/java/org/padaiyal/utilities/filesystem/FileSystemUtility.java
Comment (abprash) [reply libs/filesystem/src/main/java/org/padaiyal/utilities/filesystem/FileSystemUtility.java:339]: Had refactored to address stream leak.

Comment (abprash) [reply libs/filesystem/src/main/java/org/padaiyal/utilities/filesystem/FileSystemUtility.java:339]: @sonatype-lift ignore

File: libs/filesystem/src/main/java/org/padaiyal/utilities/filesystem/FileSystemUtility.java
Comment (abprash) [reply libs/filesystem/src/main/java/org/padaiyal/utilities/filesystem/FileSystemUtility.java]: @sonatype-lift ignore

File: libs/filesystem/src/main/java/org/padaiyal/utilities/filesystem/FileSystemUtility.java
Comment (abprash) [reply libs/filesystem/src/main/java/org/padaiyal/utilities/filesystem/FileSystemUtility.java]: @sonatype-lift ignore

**Comment by abprash** [comment]
@the-c0d3br34k3r  - the PR is done, i don't see a way to suppress the sonatype warnings we've ignored. It still seems to be considering it as a violation here. 
https://lift.sonatype.com/results/github.com/padaiyal/jMonocle/01GTZTWZK08AK669ZWZCX3RTBJ/?ph=PhaseIntroduced

File: libs/filesystem/README.md
Comment (the-c0d3br34k3r) [inline libs/filesystem/README.md:24]: Point it to a specific version rather than the generic link which leads to the maven central search page.
Refer the command line module README.

**Review by the-c0d3br34k3r — CHANGES_REQUESTED** [review]
Some of the previously requested changes are pending.
Also have requested some changes to the README. Please refer the README in the other modules.
README title and description need to be updated.
We need to have the same badges as in the other module readies only.
Please use link aliases rather than specifying the link in the actual markdown tag. Makes it easy to maintain.

**Comment by the-c0d3br34k3r** [comment]
> Some of the previously requested changes are pending.
> Also have requested some changes to the README. Please refer the README in the other modules.
> README title and description need to be updated.
> We need to have the same badges as in the other module readies only.
> Please use link aliases rather than specifying the link in the actual markdown tag. Makes it easy to maintain.

Also, the commits need to be squashed.

**Review by the-c0d3br34k3r — CHANGES_REQUESTED** [review]
Btw, the FileSystemUtilityTest is missing.

**Comment by abprash** [comment]
hi @the-c0d3br34k3r , could you take a look when possible ? I'm seeing test failures.
https://github.com/padaiyal/jMonocle/actions/runs/4614602254/jobs/8157701365?pr=11

**Comment by abprash** [comment]
Raw logs - https://pipelines.actions.githubusercontent.com/serviceHosts/c47b896d-e4ce-4b14-b53c-f87b63da89cd/_apis/pipelines/1/runs/61/signedlogcontent/3?urlExpires=2023-05-17T07%3A54%3A52.0317457Z&urlSigningMethod=HMACV1&urlSignature=MgZKPwoFthVrxJlwaYSAdIXjYxU3e8VutlUhiB20uMY%3D

**Comment by the-c0d3br34k3r** [comment]
> hi @the-c0d3br34k3r , could you take a look when possible ? I'm seeing test failures.
> https://github.com/padaiyal/jMonocle/actions/runs/4614602254/jobs/8157701365?pr=11

Hello @abprash ,
Looks like the failures are due to checkstyle/linting violations.
`Error:  Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.1:check (validate) on project filesystem: You have 539 Checkstyle violations. -> [Help 1]`

The points of failure are mentioned in the logs.

**Comment by abprash** [comment]
I had fixed the checkstyle violations. If you see the latest run, it'll be because of test failures @the-c0d3br34k3r

File: libs/filesystem/README.md
Comment (abprash) [reply libs/filesystem/README.md:104]: makes sense, will update.
LICENSE is already present.

Comment (abprash) [reply libs/filesystem/README.md:104]: Also, looks like the LICENSE files are the same for all modules. 
Do you see it changing in the future per module, or want to keep it common for the whole project ?

File: libs/filesystem/pom.xml
Comment (abprash) [reply libs/filesystem/pom.xml:18]: Got it. I can update all of them once the approval is in. Looks like they're set to 2023.02.07 in many places.
Can we set a property in parent pom and reference it in all the modules ?

File: libs/properties/src/main/java/org/padaiyal/utilities/PropertyUtility.java
Comment (abprash) [reply libs/properties/src/main/java/org/padaiyal/utilities/PropertyUtility.java]: Reverted this. Not an issue.

File: libs/filesystem/pom.xml
Comment (abprash) [reply libs/filesystem/pom.xml:18]: Got it, i can update once the PR is approved.
Keeping it open until then.

**Comment by abprash** [comment]
> I haven't investigated the root cause yet, but wanted to check with you in case they're familiar.

Fixed these. Not an issue.

File: libs/filesystem/README.md
Comment (abprash) [reply libs/filesystem/README.md:90]: Done

File: libs/filesystem/README.md
Comment (abprash) [reply libs/filesystem/README.md:75]: done

File: libs/filesystem/README.md
Comment (abprash) [reply libs/filesystem/README.md:61]: 1. So, we will not be needing the first step right ?
2. Also, the most up-to-date version will be decided once the PR is approved ?

File: libs/filesystem/README.md
Comment (abprash) [reply libs/filesystem/README.md]: done.

**Comment by abprash** [comment]
> Some of the previously requested changes are pending. Also have requested some changes to the README. Please refer the README in the other modules. README title and description need to be updated. We need to have the same badges as in the other module readies only. Please use link aliases rather than specifying the link in the actual markdown tag. Makes it easy to maintain.

Have addressed them.

**Comment by abprash** [comment]
hi @the-c0d3br34k3r 
- The Readme updates have been done.  

Pending/Open questions: 
You'd commented on the `<maven.compiler.target>14</maven.compiler.target>`, i don't see any action to be done here, until PR is approved.

Once approved:
* I can squash commits. It can be reapproved.
* I can update latest version in poms to date of approval. Then it should be good to merge.

File: libs/filesystem/README.md
Comment (the-c0d3br34k3r) [reply libs/filesystem/README.md:59]: Yes, the usage section needs to be updated.
And the version can be updated once the PR is approved.

File: libs/filesystem/README.md
Comment (the-c0d3br34k3r) [reply libs/filesystem/README.md]: Don't foresee it changing in the near future.
We have kept it separate for each module, let's just continue following that.

File: libs/filesystem/src/main/java/org/padaiyal/utilities/filesystem/FileSystemUtility.java
Comment (the-c0d3br34k3r) [reply libs/filesystem/src/main/java/org/padaiyal/utilities/filesystem/FileSystemUtility.java:303]: @sonatype-lift ignore

Comment (the-c0d3br34k3r) [reply libs/filesystem/src/main/java/org/padaiyal/utilities/filesystem/FileSystemUtility.java:303]: I've reviewed all the existing findings in Sonatype, and they just seem to be nitpicks. So ignoring them.
https://lift.sonatype.com/results/github.com/padaiyal/jMonocle/01H4VZ1QE0N5EKZRZYTHGBDFMC/?ph=PhaseIntroduced

@sonatype-lift ignoreall

Comment (the-c0d3br34k3r) [reply libs/filesystem/src/main/java/org/padaiyal/utilities/filesystem/FileSystemUtility.java:303]: @sonatype-lift ignoreall

**Review by the-c0d3br34k3r — APPROVED** [review]
@abprash LGTM.
Please make the following changes, squash and merge the PR.
1. Update the Usage section in the README to not contain the <scope> and <systemPath> tag and to also contain the latest version.
2. Update the versions in all the pom.xml files.

File: libs/filesystem/pom.xml
Comment (abprash) [reply libs/filesystem/pom.xml:18]: have updated to the date of approval.

**Comment by abprash** [comment]
> @abprash LGTM. Please make the following changes, squash and merge the PR.
> 
> 1. Update the Usage section in the README to not contain the  and  tag and to also contain the latest version.
> 2. Update the versions in all the pom.xml files.

With the first point. I've updated the latest version as the date, but not clear what you meant by `Usage section in the README to not contain the  and  tag`

**Comment by the-c0d3br34k3r** [comment]
> > @abprash LGTM. Please make the following changes, squash and merge the PR.
> > 
> > 1. Update the Usage section in the README to not contain the  and  tag and to also contain the latest version.
> > 2. Update the versions in all the pom.xml files.
> 
> With the first point. I've updated the latest version as the date, but not clear what you meant by `Usage section in the README to not contain the and tag`

Sorry, I missed it in my previous comment.
I meant removing the <scope> and <system> tags in the Usage section of the README- https://github.com/abprash/jMonocle/blob/main/libs/filesystem/README.md#usage

**Pull request merged**
